File: server_api/models/productModel.js
import { execute } from "../config/db.js";
import baseUrl from '../baseUrl.js';


export default class productModel{
    static async allproduct(){
        try{
            const [row] = await execute("SELECT * FROM products");
            return row;
        }catch(error){
            throw new Error("Database query failed: " + error.message);
        }
    }
    static async findProductById(id){
        try{
            const queryProduct = `
                SELECT p.*, b.name as brand_name 
                FROM products p
                LEFT JOIN brands b ON p.brand_id = b.id
                WHERE p.id = ?
            `;
            const [productRows] = await execute(queryProduct, [id]);
            
            if (productRows.length === 0) return null;
            const product = productRows[0];

            // Bước 2: Lấy các phiên bản (Màu sắc, bộ nhớ, giá)
            const queryVariants = `SELECT *, CONCAT(?, "/uploads/", pv.image_url) as image FROM product_variants WHERE product_id = ?`;
            const [variants] = await execute(queryVariants, [baseUrl, id]);

            // Gộp lại
            product.variants = variants; 
            return product;
        }catch(error){
            throw new Error("Lỗi lấy chi tiết sản phẩm (id): " + error.message);
        }
    }
    static async findProductByName(keyword) {
        try {
            // 1. Tách từ khóa thành mảng các từ (ví dụ: "Samsung s24" -> ["Samsung", "s24"])
            const words = keyword.trim().split(/\s+/);

            // 2. Khởi tạo câu SQL cơ bản
            // Lưu ý: WHERE 1=1 là mẹo để dễ dàng nối chuỗi AND phía sau
            let sql = `
                SELECT 
                    p.id, 
                    p.name, 
                    pv.price, 
                    p.screen_size, 
                    p.camera,
                    CONCAT(?,'/uploads/', pv.image_url) as image,
                    pv.ram,
                    pv.color
                FROM products p
                JOIN product_variants pv ON p.id = pv.product_id
                WHERE 1=1
            `;

            const params = [baseUrl]; // Tham số đầu tiên cho CONCAT

            // 3. Vòng lặp: Với mỗi từ khóa, thêm một điều kiện AND ... LIKE
            words.forEach(word => {
                sql += " AND p.name LIKE ?";
                params.push(`%${word}%`); // Thêm % vào trước và sau mỗi từ
            });

            // Nếu muốn Group By để tránh trùng lặp sản phẩm (nếu 1 sp có nhiều biến thể)
            sql += " GROUP BY p.id";

            // 4. Thực thi
            const [rows] = await execute(sql, params);
            return rows;

        } catch (error) {
            throw new Error("Lỗi tìm kiếm sản phẩm: " + error.message);
        }
    }

    static async findProductVariantById(product_variant_id){
        try{
            const [product_variants] = await execute(`
                    SELECT *, CONCAT(?, "/uploads/", pv.image_url) as image FROM product_variants pv WHERE p.id = ?
                `,[baseUrl,product_variant_id]);

            if(product_variants.length === 0) {
                throw new Error('Không tìm thấy sản phẩm này');
            }

            return product_variants[0] ?? null;
        }
        catch(error){
            throw new Error('Lỗi Lấy biến thể sản phẩm bằng id: ' + error.message)
        }
    }

    static async findProductVariantByProductID(product_id){
        try{
            const [rows] = await execute('SELECT p.brand_id, p.name, p.description, p.screen_size, p.cpu, p.camera, p.battery, pv.product_id, pv.color, pv.ram, pv.storage, pv.price, pv.stock_quantity, CONCAT(?, "/uploads/", pv.image_url) as image FROM products p, product_variants pv WHERE p.id = pv.product_id AND p.id = ?',[baseUrl,product_id]);
            if(rows.length === 0) throw new Error('Không tìm thấy product_id');

            return rows;
        }
        catch(error){
            throw new Error('Lỗi lấy biến thể sản phẩm bằng id sản phẩm: ' + error.message);
        }
    }

    static async getProductVariants(){
        try{
            // Sửa lại: CONCAT(?, '/uploads/', pv.image_url)
            const [product_variant] = await execute(
                'SELECT pv.id, p.brand_id, p.name, p.screen_size, p.camera, pv.storage, pv.product_id, pv.color, pv.ram, pv.price, pv.stock_quantity, CONCAT(?, "/uploads/", pv.image_url) as image FROM products p, product_variants pv WHERE p.id = pv.product_id', 
                [baseUrl] // Biến này sẽ điền vào dấu ? đầu tiên
            );
            if(product_variant.length === 0) return [];
            return product_variant;
        }catch(error){
            throw new Error("Lỗi lấy danh sách biến thể sản phẩm: " + error.message);
        }
    }
}

File: flutter_application_f4_mobile/lib/Model/product_model.dart

// 1. Thêm dòng import này để lấy link ngrok từ Config
import '../Config/baseUrl.dart'; // Đảm bảo bạn đã có file này chứa link Ngrok

class Product {
  final int id;
  final int brandId;
  final String name;
  final String description;
  final String screenSize;
  final String cpu;
  final String camera;
  final String battery;
  final String imageUrl;
  final int soldCount;
  final String createdAt;
  final List<ProductVariant>? variants;

  Product({
    required this.id,
    required this.brandId,
    required this.name,
    required this.description,
    required this.screenSize,
    required this.cpu,
    required this.camera,
    required this.battery,
    required this.imageUrl,
    required this.soldCount,
    required this.createdAt,
    this.variants,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    // Logic ghép link ảnh cho Product cha
    String getFullUrl(String? path) {
      if (path == null || path.isEmpty) return "";
      if (path.startsWith("http")) return path;

      // Xử lý link Ngrok: bỏ đuôi /api nếu baseUrl có chứa /api để ghép với /uploads
      String rootUrl = baseUrl.endsWith('/api') 
          ? baseUrl.replaceAll('/api', '') 
          : baseUrl;
      
      // Xử lý dấu gạch chéo
      String cleanRoot = rootUrl.endsWith('/') ? rootUrl.substring(0, rootUrl.length - 1) : rootUrl;
      String cleanPath = path.startsWith('/') ? path : '/$path';
      
      return "$cleanRoot$cleanPath";
    }

    List<ProductVariant>? variantsList;
    if (json['variants'] != null) {
      var list = json['variants'] as List;
      variantsList = list.map((i) => ProductVariant.fromJson(i)).toList();
    }

    return Product(
      id: json['id'] ?? 0,
      brandId: json['brand_id'] ?? 0,
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      screenSize: json['screen_size'] ?? '',
      cpu: json['cpu'] ?? '',
      camera: json['camera'] ?? '',
      battery: json['battery'] ?? '',
      // Tự động ghép link cho ảnh
      imageUrl: getFullUrl(json['image_url'] ?? json['imageUrl'] ?? json['image']),
      soldCount: json['sold_count'] ?? 0,
      createdAt: json['created_at'] ?? '',
      variants: variantsList,
    );
  }
}

class ProductVariant {
  final int id;
  final int? brandId;
  final String? name;
  final String? description;
  final String? screenSize;
  final String? cpu;
  final String? camera;
  final String? battery;
  final int? productId;
  final String? color;
  final String? ram;
  final String? storage;
  final double? price;
  final int? stockQuantity;
  final String? imageUrl;

  ProductVariant({
    required this.id,
    this.brandId,
    this.name,
    this.description,
    this.screenSize,
    this.cpu,
    this.camera,
    this.battery,
    this.productId,
    this.color,
    this.ram,
    this.storage,
    this.price,
    this.stockQuantity,
    this.imageUrl,
  });

  factory ProductVariant.fromJson(Map<String, dynamic> json) {
    // Logic ghép link ảnh cho ProductVariant (QUAN TRỌNG VỚI TRANG HOME)
    String getFullUrl(String? path) {
      if (path == null || path.isEmpty) return "";
      if (path.startsWith("http")) return path;

      String rootUrl = baseUrl.endsWith('/api') 
          ? baseUrl.replaceAll('/api', '') 
          : baseUrl;
          
      String cleanRoot = rootUrl.endsWith('/') ? rootUrl.substring(0, rootUrl.length - 1) : rootUrl;
      String cleanPath = path.startsWith('/') ? path : '/$path';
      
      return "$cleanRoot$cleanPath";
    }

    return ProductVariant(
      id: json['id'] ?? 0,
      brandId: json['brand_id'] != null ? int.tryParse(json['brand_id'].toString()) : null,
      name: json['name'] as String?,
      description: json['description'],
      screenSize: json['screen_size'],
      cpu: json['cpu'],
      camera: json['camera'],
      battery: json['battery'],
      productId: json['product_id'] != null ? int.tryParse(json['product_id'].toString()) : null,
      color: json['color'],
      ram: json['ram'],
      storage: json['storage'],
      price: json['price'] != null ? double.tryParse(json['price'].toString()) : 0.0,
      stockQuantity: json['stock_quantity'] != null ? int.tryParse(json['stock_quantity'].toString()) : 0,
      
      // SỬA LỖI Ở ĐÂY: Sử dụng hàm getFullUrl để đảm bảo ảnh luôn có link full
      imageUrl: getFullUrl(json['image'] ?? json['imageUrl'] ?? json['image_url']),
    );
  }
}